# **Data Cleaning and Preparation for EDA**

## Objectives

* Download data from Kaggle and load into a pandas dataframe
* Clean and prepare dataset
* Engineer features for exploratory data analysis (EDA)

## Inputs

* Student academic performance dataset from Kaggle https://www.kaggle.com/datasets/sonalshinde123/student-academic-performance-dataset 

## Outputs

* Cleaned dataset output to "academic_performance_cleaned.csv". Location: data folder in student academic performance repository 



---

# Change working directory

We need to change the working directory from its current folder to its parent folder
* We access the current directory with os.getcwd()

In [1]:
import os
current_dir = os.getcwd()
current_dir

'c:\\Users\\tb975\\OneDrive\\Documents\\vs_code_projects\\Student-Academic-Performance\\jupyter_notebooks'

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [2]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


Confirm the new current directory

In [3]:
current_dir = os.getcwd()
current_dir

'c:\\Users\\tb975\\OneDrive\\Documents\\vs_code_projects\\Student-Academic-Performance'

# Section 1: Data loading and investigation

Start by importing the relavent python libraries I will need for data cleaning and exploration.

In [4]:
#import libraries for data manipulation
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns

I will now read the csv file located in the data folder into a pandas dataframe for manipulation.

In [5]:
#load csv into pandas dataframe
df = pd.read_csv('data/student_academic_performance_raw.csv')
#display the first 5 rows
df.head()

Student_ID  Attendance (%)  Internal Test 1 (out of 40)  \
0      S1000              84                           30   
1      S1001              91                           24   
2      S1002              73                           29   
3      S1003              80                           36   
4      S1004              84                           31   

   Internal Test 2 (out of 40)  Assignment Score (out of 10)  \
0                           36                             7   
1                           38                             6   
2                           26                             7   
3                           35                             7   
4                           37                             8   

   Daily Study Hours  Final Exam Marks (out of 100)  
0                  3                             72  
1                  3                             56  
2                  3                             56  
3                  3                             74  
4                  3                             66

I can investigate the size of the dataset by printing the shape of the table. This dataset has 2000 rows and 7 columns. This represents 2000 different pupils and their academic indicators.

In [6]:
#print the shape of the dataframe
print(df.shape)

(2000, 7)


I can also assess the column names and datatypes of the dataset using .info(). In section 2 I will consider the need to change datatypes as part of the cleaning process.

In [7]:
#print column names and datatypes 
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 7 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   Student_ID                     2000 non-null   object
 1   Attendance (%)                 2000 non-null   int64 
 2   Internal Test 1 (out of 40)    2000 non-null   int64 
 3   Internal Test 2 (out of 40)    2000 non-null   int64 
 4   Assignment Score (out of 10)   2000 non-null   int64 
 5   Daily Study Hours              2000 non-null   int64 
 6   Final Exam Marks (out of 100)  2000 non-null   int64 
dtypes: int64(6), object(1)
memory usage: 109.5+ KB
None


From the information above it looks like there are no null values. This can be confirmed below by summing the null values. Zero found in each column.

In [8]:
#check for null values and sum for each column
df.isna().sum()

Student_ID                       0
Attendance (%)                   0
Internal Test 1 (out of 40)      0
Internal Test 2 (out of 40)      0
Assignment Score (out of 10)     0
Daily Study Hours                0
Final Exam Marks (out of 100)    0
dtype: int64

I can check for duplicate rows also: zero found

In [9]:
#check for duplicates sum results
df.duplicated().sum()

0

The describe module give me basic summary statistics for the numerical columns. I can assess the mean, standard deviation, maximum, and minimum values. 

These summary statistics can be useful as a first pass to check the values make sense. Checking the maximum and minumum values tells us the following:
- attendence has no values above 100% or below 0%
- both test scores (out of 40) do not have values above 40 or below 0
- the assignment score (out of 10) does not have any vlaues above 10 or below 0 
- daily study hours range from 1 to 5, which seams reasonable
- final exam marks (out of 100) have no values above 100 or below 0 

In [10]:
#get summary statistics round to 2d.p.
df.describe().round(2)

Attendance (%)  Internal Test 1 (out of 40)  \
count         2000.00                      2000.00   
mean            84.89                        32.12   
std              7.76                         4.56   
min             52.00                        18.00   
25%             80.00                        29.00   
50%             85.00                        32.00   
75%             90.00                        35.00   
max            100.00                        40.00   

       Internal Test 2 (out of 40)  Assignment Score (out of 10)  \
count                      2000.00                       2000.00   
mean                         32.46                          7.51   
std                           4.52                          1.02   
min                          16.00                          4.00   
25%                          29.00                          7.00   
50%                          33.00                          8.00   
75%                          36.00                          8.00   
max                          40.00                         10.00   

       Daily Study Hours  Final Exam Marks (out of 100)  
count            2000.00                        2000.00  
mean                2.82                          64.86  
std                 0.61                          11.34  
min                 1.00                          25.00  
25%                 2.00                          58.00  
50%                 3.00                          65.00  
75%                 3.00                          73.00  
max                 5.00                         100.00

---

# Section 2: Data cleaning

Quick reminder of the data

In [11]:
#show the first 5 rows of the data
df.head()

Student_ID  Attendance (%)  Internal Test 1 (out of 40)  \
0      S1000              84                           30   
1      S1001              91                           24   
2      S1002              73                           29   
3      S1003              80                           36   
4      S1004              84                           31   

   Internal Test 2 (out of 40)  Assignment Score (out of 10)  \
0                           36                             7   
1                           38                             6   
2                           26                             7   
3                           35                             7   
4                           37                             8   

   Daily Study Hours  Final Exam Marks (out of 100)  
0                  3                             72  
1                  3                             56  
2                  3                             56  
3                  3                             74  
4                  3                             66

An ethical consideration to make when using this dataset is the privacy of the pupils and with regard to them being identified by their student ID. I do not need studentID for my analysis and hence can drop this column. This means that the data is fully annonomysed and no pupil can be identified in this dataset. 

Considering the privacy of participants in any dataset is an important ethical consideration. This dataset is synthetic, meaning that the rows do not represent real pupil, but I will remove the studentID column anyway to simulate what would be best practice in the real world. 

In [12]:
#drop student ID column
df = df.drop('Student_ID', axis=1)
df

Attendance (%)  Internal Test 1 (out of 40)  \
0                 84                           30   
1                 91                           24   
2                 73                           29   
3                 80                           36   
4                 84                           31   
...              ...                          ...   
1995              82                           31   
1996              78                           38   
1997              78                           30   
1998              82                           29   
1999              97                           32   

      Internal Test 2 (out of 40)  Assignment Score (out of 10)  \
0                              36                             7   
1                              38                             6   
2                              26                             7   
3                              35                             7   
4                              37                             8   
...                           ...                           ...   
1995                           28                             6   
1996                           27                             7   
1997                           33                             9   
1998                           40                             8   
1999                           38                             7   

      Daily Study Hours  Final Exam Marks (out of 100)  
0                     3                             72  
1                     3                             56  
2                     3                             56  
3                     3                             74  
4                     3                             66  
...                 ...                            ...  
1995                  2                             52  
1996                  2                             57  
1997                  2                             61  
1998                  3                             59  
1999                  3                             64  

[2000 rows x 6 columns]

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 6 columns):
 #   Column                         Non-Null Count  Dtype
---  ------                         --------------  -----
 0   Attendance (%)                 2000 non-null   int64
 1   Internal Test 1 (out of 40)    2000 non-null   int64
 2   Internal Test 2 (out of 40)    2000 non-null   int64
 3   Assignment Score (out of 10)   2000 non-null   int64
 4   Daily Study Hours              2000 non-null   int64
 5   Final Exam Marks (out of 100)  2000 non-null   int64
dtypes: int64(6)
memory usage: 93.9 KB


Now there are 6 columns left, all with a datatype of int64. In each column the largest number is 100. Because of this we can make the memory useage of this dataset smaller by changing their datatypes to int8. This provides a smaller memory allocation for each number. This dataset is relatively small, so the impact of doing this is not huge, but with large datasets then saving memory may make a significant difference to performance. 

In [14]:
#save the column names in a list called cols
cols = df.columns

#loop through the colums in cols and change the datatype from int64 to int8

df = df.astype({col: 'int8' for col in cols})

#display new datatypes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 6 columns):
 #   Column                         Non-Null Count  Dtype
---  ------                         --------------  -----
 0   Attendance (%)                 2000 non-null   int8 
 1   Internal Test 1 (out of 40)    2000 non-null   int8 
 2   Internal Test 2 (out of 40)    2000 non-null   int8 
 3   Assignment Score (out of 10)   2000 non-null   int8 
 4   Daily Study Hours              2000 non-null   int8 
 5   Final Exam Marks (out of 100)  2000 non-null   int8 
dtypes: int8(6)
memory usage: 11.8 KB


Memory useage has decreased from 93.9kB to 11.8kB

## Feature Engineering

This dataset contains two sets of internal test results. I have decided to average these test results in a new column. This means that when it comes to modelling the dataset their performance in internal tests is represented once (by the average) not twice (by both columns). The hope is that this will avoid redundancy and redice multicollinearity.

The new column will be the mean average of the two internal test results for each pupil. The mean represents a central value and it calculated by summing the test results for each pupil and dividing them by how many tests there are (in this case 2). 

This combines the two internal assessment variables (Test 1 and Test 2) into a single feature representing mean internal performance. The original variables were retained in the dataset for transparency.

In [15]:
#make a new column called average test score, which is the mean of the two internal test scores
df['Average Test Score'] = (df['Internal Test 1 (out of 40)'] + df['Internal Test 2 (out of 40)']) /2
#display top ten rows
df.head(10)

Attendance (%)  Internal Test 1 (out of 40)  Internal Test 2 (out of 40)  \
0              84                           30                           36   
1              91                           24                           38   
2              73                           29                           26   
3              80                           36                           35   
4              84                           31                           37   
5             100                           34                           34   
6              96                           40                           36   
7              83                           39                           37   
8              91                           30                           37   
9              87                           27                           37   

   Assignment Score (out of 10)  Daily Study Hours  \
0                             7                  3   
1                             6                  3   
2                             7                  3   
3                             7                  3   
4                             8                  3   
5                             7                  3   
6                             8                  3   
7                             7                  3   
8                             8                  2   
9                             8                  3   

   Final Exam Marks (out of 100)  Average Test Score  
0                             72                33.0  
1                             56                31.0  
2                             56                27.5  
3                             74                35.5  
4                             66                34.0  
5                             79                34.0  
6                             83                38.0  
7                             77                38.0  
8                             71                33.5  
9                             61                32.0

Hypothesis 3 compares splits pupils into a high study hours group and a low study hours group to assess if studyng for a longer time has a significant impact on final exam marks.

I will use the daily study hours column to split pupils into two categories, high study hours and low study hours. Pupils with above a certain threshold number of hours will be given the category high study hours and vice versa.

To decide what the high/low study hours threshold will be I will investigate the distribution of daily study hours.

In [16]:
#show me how many pupils are studying each number of hours, sort from smallest largest number of hours
df['Daily Study Hours'].value_counts().sort_index()

Daily Study Hours
1      14
2     533
3    1248
4     202
5       3
Name: count, dtype: int64

I am interested in the median study hours value

In [17]:
#calculate the median and print
median = df['Daily Study Hours'].median()
print(median)

3.0


The median is 3 hours of study per day, and the vast majority of pupils (1248) are studying for 3 hours. 1 hour of study is rare and 5 hours is very rare. I am going to say that a typical number of hours to be studying is 3, so I will define low study effort as 3 or fewer hours. Therefore more than 3 hours represents deliberate extra effort so is defined as a high number of study hours. 

In [18]:
def study_groups (hours):
    '''
    Function categorises hours less than or equal to 3 into low study hours
    and more than 3 into high study hours
    '''
    if hours <= 3:
        hours = 'low'
    elif hours > 3:
        hours = 'high'
    return hours

df['Study Group'] = df['Daily Study Hours'].apply(study_groups)
df

Attendance (%)  Internal Test 1 (out of 40)  \
0                 84                           30   
1                 91                           24   
2                 73                           29   
3                 80                           36   
4                 84                           31   
...              ...                          ...   
1995              82                           31   
1996              78                           38   
1997              78                           30   
1998              82                           29   
1999              97                           32   

      Internal Test 2 (out of 40)  Assignment Score (out of 10)  \
0                              36                             7   
1                              38                             6   
2                              26                             7   
3                              35                             7   
4                              37                             8   
...                           ...                           ...   
1995                           28                             6   
1996                           27                             7   
1997                           33                             9   
1998                           40                             8   
1999                           38                             7   

      Daily Study Hours  Final Exam Marks (out of 100)  Average Test Score  \
0                     3                             72                33.0   
1                     3                             56                31.0   
2                     3                             56                27.5   
3                     3                             74                35.5   
4                     3                             66                34.0   
...                 ...                            ...                 ...   
1995                  2                             52                29.5   
1996                  2                             57                32.5   
1997                  2                             61                31.5   
1998                  3                             59                34.5   
1999                  3                             64                35.0   

     Study Group  
0            low  
1            low  
2            low  
3            low  
4            low  
...          ...  
1995         low  
1996         low  
1997         low  
1998         low  
1999         low  

[2000 rows x 8 columns]

---

I now have a cleaned dataset with all the features required for exploratory data analysis.

# Push files to Repo

In [ ]:
# save files in a new cleaned csv
df.to_csv('data/academic_performance_cleaned.csv', index=False)